### ISSUE TITLE
Adding Metadata to video  frame i.e adding sensor data and any custom metadata 

### ISSUE THREAD

**Issue by muni915** [pedroSG94/rtmp-rtsp-stream-client-java#196]
Hi Pedro  currently you are sending video and audio streaming by muxing video encoder and audio encoder to media mux along with video and audio track how to send the metadata track I. E any sensor data for each frame proximity sensor data to the server I am required to stream proximity sensor data continuously along with video streaming synchronsly to server how we can achieve sending metadata to server continuesly

**Comment by pedroSG94** [comment]
No, mediamuxer is used to record mp4 video. If you want send some info to server you can try modify rtp packets and add some info in that packets.
RTSP:
Video:
https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/blob/master/rtsp/src/main/java/com/pedro/rtsp/rtp/packets/H264Packet.java
Audio:
https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/blob/master/rtsp/src/main/java/com/pedro/rtsp/rtp/packets/AacPacket.java
RTMP:
https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/blob/master/rtmp/src/main/java/net/ossrs/rtmp/SrsFlvMuxer.java

**Comment by muni915** [comment]
hi pedro here i want send timed matadata along with video frame i.e each video frame contains some different sensor motion event i.e for 30frames of video it will contain 30 sensor motion values into the ByteBuffer it will broadcast to server so can you please give me a detailed example to add metadata to each video frame in the above example i am not able to understand how to add metadata
and  i am following an link in where adding the metadata [https://developer.android.com/reference/android/media/MediaMuxer](url) 
package com.pedro.rtplibrary.base.camera1base  `
  @Override
  public void getH264Data(ByteBuffer h264Buffer, MediaCodec.BufferInfo info) {
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR2 && recording) {
      if (info.flags == MediaCodec.BUFFER_FLAG_KEY_FRAME
          && !canRecord
          && videoFormat != null
          && audioFormat != null) {
        videoTrack = mediaMuxer.addTrack(videoFormat);
        audioTrack = mediaMuxer.addTrack(audioFormat);


        //adding the metadata task

     //     metadataTrackIndex = mediaMuxer.addTrack(metadataFormat);
          Log.d(TAG,"metadataTrackIndex"+metadataTrackIndex);

        // Setup Metadata Track

         /* try {
           //   MediaCodec mediaCodec=   MediaCodec.createEncoderByType("application/gyro");


           metadataFormat = new MediaFormat();
              metadataFormat.setString(KEY_MIME, "application/gyro");
          //    mediaCodec.configure(metadataFormat,null,null,MediaCodec.CONFIGURE_FLAG_ENCODE);

            Log.d("camera1Base-","mediacodeinfo name"+mediaCodec.getCodecInfo().getName());

              Log.d("camera1Base-","media name supported types-"+mediaCodec.getCodecInfo().getSupportedTypes());


          } catch (IOException e) {
              e.printStackTrace();
          }*/






        // Allocate bytebuffer and write gyro data(x,y,z) into it.
        metaData = ByteBuffer.allocate(h264Buffer.capacity());
        metaData.putFloat(101);
        metaData.putFloat(102);
        metaData.putFloat(103);



        metaInfo = new MediaCodec.BufferInfo();
        // Associate this metadata with the video frame by setting
        // the same timestamp as the video frame.
        metaInfo.presentationTimeUs = info.presentationTimeUs;
        metaInfo.offset = info.offset;
        metaInfo.flags = info.flags;
        metaInfo.size = info.size;



        Log.d(TAG,"getH264Data-videoTrack"+videoTrack);

        Log.d(TAG,"getH264Data-audioTrack"+audioTrack);

       Log.d(TAG,"getH264Data-metadatrack"+metadataTrackIndex);


        canRecord = true;
     //     mediaMuxer.setLocation(30,30);


          Log.d(TAG, "Frame (" + "--" + ") " +
                  "PresentationTimeUs:" + info.presentationTimeUs +
                  " Flags:" + info.flags +
                  " TrackIndex:" + videoTrack +
                  " Size(KB) " + info.size / 1024);
      }
      if (canRecord)
      {
      mediaMuxer.writeSampleData(videoTrack, h264Buffer, info);
      Log.d(TAG,"metaInfo-presentationTimeUs"+metaInfo.presentationTimeUs+"\n metaData-bytebuffer"+metaData.getFloat(0));
 //   mediaMuxer.writeSampleData(metadataTrackIndex, metaData, metaInfo);
      }
    }
    if (streaming) getH264DataRtp(h264Buffer, info);
  }`

i tried here but i am getting error app crashes so pedro give me a detailed example where i can add metadata to video frame then your app will becom more productive which handles audio and video and metadata three categerious will suport

**Comment by pedroSG94** [comment]
You can't send metadata using MediaMuxer class because in my library it is used to record a video in the local storage for that I link you video packets and audio packets where you can modify it.

The problem here is that by default you can do it in RTSP or RTMP protocol (or I don't know how do it so I can't give you an example) so you need custom your server to receive a buffer that is not correct (because your buffer contains a value that is not valid in the stream protocol in your case sensor values).
How can you do it?
For example you can:
- Modify buffer in class that I link you to add the idea is add to packet some bytes that contains this info.
- Now, in the server. You know where is this info in the packet (at first 5 bytes are your sensor values for example) so you can extract it and use valid bytes (the original video frame).

**Comment by muni915** [comment]
Yes pedro we can add the bytes information to video frames other wise i am
thinking that another way is we can send DYNAMIC rtsp headers so in this
headers we can stream sensor data to the server while it listen to the
sensor motions so can you give me example how to send the dynamic headers
added to the rtsp/rtmp protocal which will inlude timestamp and sensor
events so can you please tell me with a example how to send the dynamic
header information to server so that the server will detect the frames
according to the header thanking you pedro

**Comment by pedroSG94** [comment]
I haven't an example of it so someone need develop it. Your case is so specific and no useful for others users so I will not develop it.
